### PR TITLE
support converged charging

### PR DESCRIPTION
**Pull request by Roy-Hu**
1. PDU-based Charging: SMF collects charging information related to data volumes sent to and received by the
UE per user per PDU session (achieved by flow-based charging with a specific rating group)
2. Flow-Based Charging: SMF categorizes the service data flows within PDU session data traffic by rating group and/or
combination of the rating group
3. The PCC will define the flow's charging method(online/offline) and its rating group. Set the charging of default PCC (FlowDescription: "permit out ip from any to assigned" )to PDU-based Charging. All the data paths created for pcc need to additionally apply the URR created for this default pcc for recording all volumes in a PDU session.

### CODE REVIEW

**Comment by tim-ywliu** [comment]
@Roy-Hu @Tai-shang-chun
I rebased this PR and updated openapi/pfcp hash to the latest, but it will go test fail.
Please help fix it. Thanks.

File: internal/context/sm_context_policy.go [hunk 1]
```diff
@@ -68,6 +68,54 @@ func (c *SMContext) RemoveQosFlow(qfi uint8) {
...
+// For urr that created for Pdu session level charging, it show be applied to all data path
+func (c *SMContext) addPduLevelChargingRuleToFlow(pccRules map[string]*PCCRule) {
+	var pduLevelChargingUrrs []*URR
+	var pccIdsForPduSession []string
+
+	for id, pcc := range pccRules {
+		if chargingLevel, err := pcc.IdentifyChargingLevel(); err != nil {
+			continue
+		} else if chargingLevel == PduSessionCharging {
+			pccIdsForPduSession = append(pccIdsForPduSession, id)
```
Comment (ianchen0119) [inline internal/context/sm_context_policy.go:80]: I think `pccIdsForPduSession` is redundant, this part can be changed to:
```go
pduPcc := pccRules[id]
pduLevelChargingUrrs = pduPcc.Datapath.GetChargingUrr(c)
```

File: internal/context/sm_context_policy.go [hunk 2]
```diff
@@ -68,6 +68,54 @@ func (c *SMContext) RemoveQosFlow(qfi uint8) {
...
+		}
+	}
+
+	for _, pduPccid := range pccIdsForPduSession {
```
Comment (ianchen0119) [inline internal/context/sm_context_policy.go:84]: remove this part.

File: internal/sbi/producer/callback.go [hunk 3]
```diff
@@ -11,6 +12,58 @@ import (
...
+
+						upfId := smContext.ChargingInfo[urr.URRID].UpfId
+						upfUrrMap[upfId] = append(upfUrrMap[upfId], urr)
+					}
+				}
+
+			}
+
+			for upfId, urrList := range upfUrrMap {
+				upf := smf_context.GetUpfById(upfId)
```
Comment (ianchen0119) [inline internal/sbi/producer/callback.go:50]: Should do nil checking

File: internal/sbi/producer/charging_trigger.go [hunk 4]
```diff
@@ -0,0 +1,351 @@
...
+				}
+			}
+
+			if len(upfUrrMap) == 0 {
+				logger.ChargingLog.Infof("Do not have urr that need to update charging information")
+				return
+			}
+
+			for upfId, urrList := range upfUrrMap {
+				upf := smf_context.GetUpfById(upfId)
```
Comment (ianchen0119) [inline internal/sbi/producer/charging_trigger.go:111]: Should do nil checking

File: internal/sbi/producer/charging_trigger.go [hunk 5]
```diff
@@ -0,0 +1,351 @@
...
+							smContext.Log.Infof("Add Volume Limit Expiry Timer for Pdu session, it's rating group is [%d]", rg)
+
+							if chgInfo.VolumeLimitExpiryTimer != nil {
+								chgInfo.VolumeLimitExpiryTimer.Stop()
+								chgInfo.VolumeLimitExpiryTimer = nil
+							}
+
+							chgInfo.VolumeLimitExpiryTimer = smf_context.NewTimer(time.Duration(t.VolumeLimit)*time.Second, 1, func(expireTimes int32) {
+								urrList := []*smf_context.URR{urr}
+								upf := smf_context.GetUpfById(ui.UPFID)
```
Comment (ianchen0119) [inline internal/sbi/producer/charging_trigger.go:275]: Should do nil checking

File: internal/sbi/producer/charging_trigger.go [hunk 6]
```diff
@@ -0,0 +1,351 @@
...
+							smContext.Log.Infof("Add Volume Limit Expiry Timer for rating group [%d] ", rg)
+
+							if chgInfo.VolumeLimitExpiryTimer != nil {
+								chgInfo.VolumeLimitExpiryTimer.Stop()
+								chgInfo.VolumeLimitExpiryTimer = nil
+							}
+
+							chgInfo.VolumeLimitExpiryTimer = smf_context.NewTimer(time.Duration(t.VolumeLimit)*time.Second, 1, func(expireTimes int32) {
+								urrList := []*smf_context.URR{urr}
+								upf := smf_context.GetUpfById(ui.UPFID)
```
Comment (ianchen0119) [inline internal/sbi/producer/charging_trigger.go:295]: Should do nil checking

File: internal/sbi/producer/charging_trigger.go [hunk 7]
```diff
@@ -0,0 +1,351 @@
...
+								chgInfo.VolumeLimitExpiryTimer = nil
+							})
+						}
+					}
+				case models.TriggerType_MAX_NUMBER_OF_CHANGES_IN_CHARGING_CONDITIONS:
+					switch chgInfo.ChargingLevel {
+					case smf_context.PduSessionCharging:
+						chgInfo.EventLimitExpiryTimer = smf_context.NewTimer(time.Duration(t.EventLimit)*time.Second, 1, func(expireTimes int32) {
+							urrList := []*smf_context.URR{urr}
+							upf := smf_context.GetUpfById(ui.UPFID)
```
Comment (ianchen0119) [inline internal/sbi/producer/charging_trigger.go:309]: Should do nil checking

File: internal/pfcp/handler/handler.go [hunk 8]
```diff
@@ -190,54 +191,16 @@ func HandlePfcpSessionReportRequest(msg *pfcpUdp.Message) {
 	}
 
 	if req.ReportType.Usar && req.UsageReport != nil {
-		HandleReports(req.UsageReport, nil, nil, smContext, upfNodeID)
+		smContext.HandleReports(req.UsageReport, nil, nil, upfNodeID, "")
+		// After recieving the Usage Report, it should send charging request to the CHF
+		// and update the URR with the quota or other charging information  according to
+		// the charging response
+		go func() {
+			producer.ReportUsageAndUpdateQuota(smContext)
```
Comment (ianchen0119) [inline internal/pfcp/handler/handler.go:199]: Do we need additional goroutine for updating the information?

File: internal/pfcp/message/build.go [hunk 9]
```diff
@@ -427,6 +500,7 @@ func BuildPfcpSessionModificationRequest(
 
 	msg.UpdatePDR = make([]*pfcp.UpdatePDR, 0, 2)
 	msg.UpdateFAR = make([]*pfcp.UpdateFAR, 0, 2)
+	msg.UpdateURR = make([]*pfcp.UpdateURR, 0, 12)
```
Comment (ianchen0119) [inline internal/pfcp/message/build.go:503]:
```go
msg.UpdateURR = make([]*pfcp.UpdateURR, 0, len(urrList))
```

File: internal/sbi/producer/callback.go [hunk 10]
```diff
@@ -11,6 +12,58 @@ import (
...
+		return httpwrapper.NewResponse(int(problemDetails.Status), nil, problemDetails)
+	} else {
+		return httpwrapper.NewResponse(http.StatusNoContent, nil, nil)
+	}
+}
+
+// While recieve Charging Notification from CHF, SMF will send Charging Information to CHF and update UPF
+func chargingNotificationProcedure(req models.ChargingNotifyRequest, smContextRef string) *models.ProblemDetails {
+	if smContext := smf_context.GetSMContextByRef(smContextRef); smContext != nil {
+		go func() {
```
Comment (ianchen0119) [inline internal/sbi/producer/callback.go:29]: Do we need additional goroutine here?

File: internal/sbi/producer/pdu_session.go [hunk 11]
```diff
@@ -181,6 +181,15 @@ func HandlePDUSessionSMContextCreate(request models.PostSmContextsRequest) *http
...
 	smContext.SMPolicyID = smPolicyID
 
+	// PDU　session create is a charging event
+	logger.PduSessLog.Infof("CHF Selection for SMContext SUPI[%s] PDUSessionID[%d]\n",
+		smContext.Supi, smContext.PDUSessionID)
+	if err := smContext.CHFSelection(); err != nil {
+		logger.PduSessLog.Errorln("chf selection error:", err)
+	}
+
+	CreateChargingSession(smContext)
```
Comment (ianchen0119) [inline internal/sbi/producer/pdu_session.go:191]:
```go
	// PDU　session create is a charging event
	logger.PduSessLog.Infof("CHF Selection for SMContext SUPI[%s] PDUSessionID[%d]\n",
		smContext.Supi, smContext.PDUSessionID)
	if err := smContext.CHFSelection(); err != nil {
		logger.PduSessLog.Errorln("chf selection error:", err)
	} else {
        CreateChargingSession(smContext)
    }
```

File: internal/sbi/producer/pdu_session.go [hunk 12]
```diff
@@ -975,6 +1010,7 @@ func HandlePDUSessionSMContextLocalRelease(smContext *smf_context.SMContext, cre
 }
 
 func releaseSession(smContext *smf_context.SMContext) smf_context.PFCPSessionResponseStatus {
+
```
Comment (ianchen0119) [inline internal/sbi/producer/pdu_session.go:1013]: remove this line.

**Pull request closed without merging**